mmdet/utils/memory.py

```diff
@@ -0,0 +1,217 @@
+                                 'cannot get dtype and device.')
+
+            # Try to use FP16
+            fp16_args = convert_tensor_type(args, dst_type=torch.half)
+            fp16_kwargs = convert_tensor_type(kwargs, dst_type=torch.half)
+            self.logger.info(f'Attempting to copy inputs of {str(func)} '
+                             f'to fp16 due to CUDA OOM')
+
+            if self.keep_type:
+                # get input tensor type, the output type will same as
+                # the first parameter type.
+                with _ignore_torch_cuda_oom():
+                    self.logger.info(f'Trying to convert output to {dtype}')
+                    output = func(*fp16_args, **fp16_kwargs)
+                    output = convert_tensor_type(
+                        output, src_type=torch.half, dst_type=dtype)
+                    return output
+                self.logger.info(f'Cannot convert output to {dtype} due to '
+                                 'CUDA OOM.')
+
+            if self.return_fp16:
+                with _ignore_torch_cuda_oom():
+                    # try to convert outputs to fp16
+                    self.logger.info('Trying to convert outputs to fp16')
+                    return func(*fp16_args, **fp16_kwargs)
+                warnings.warn('Cannot convert outputs to fp16 due to CUDA OOM')
+                self.logger.info('Cannot convert output to fp16 due to '
+                                 'CUDA OOM.')
+
+            # Try on CPU. This slows down the code significantly,
+            # therefore print a notice.
+            if self.convert_cpu:
+                self.logger.info(f'Attempting to copy inputs of {str(func)} '
+                                 f'to CPU due to CUDA OOM')
+                cpu_device = torch.empty(0).device
+                cpu_args = convert_tensor_type(args, dst_type=cpu_device)
+                cpu_kwargs = convert_tensor_type(kwargs, dst_type=cpu_device)
+                if self.return_gpu:
+                    # try to convert outputs to GPU
+                    with _ignore_torch_cuda_oom():
+                        self.logger.info(f'Convert outputs to GPU '
+                                         f'(device={device})')
+                        output = func(*cpu_args, **cpu_kwargs)
+                        output = convert_tensor_type(
+                            output, src_type=cpu_device, dst_type=device)
+                        return output
+
+                warnings.warn('Cannot convert output to GPU due to CUDA OOM, '
+                              'the output is now on CPU, which might cause '
+                              'errors if the output need to interact with GPU '
+                              'data in subsequent operations')
+                self.logger.info('Cannot convert output to GPU due to '
+                                 'CUDA OOM, the output is on CPU now.')
+
+                return func(*cpu_args, **cpu_kwargs)
+            else:
```

Comment:
The dtype of output tensor fp16 when  `keep_type=False, return_fp16=False, convert_cpu=False`, how to handle with this case?

Reply:
It will re-try to use fp16 as input by using `func(*fp16_args, **fp16_kwargs)`.

Reply:
Shall we add this description for this case (The dtype of output tensor is fp16 when keep_type=False, return_fp16=False, convert_cpu=False)?